python311Packages.babeltrace: init at 1.5.11

wentasah opened the pull request:
## Description of changes

Rebase of #285681 plus additional fixes.

In the original PR, @hacker1024 (who [agrees with this PR](https://github.com/NixOS/nixpkgs/pull/285681#issuecomment-2106054084))  says:
> This PR adds the [Babeltrace 2](https://github.com/efficios/babeltrace) package along with its Python bindings. It is ["completely independent from Babeltrace 1"](https://github.com/efficios/babeltrace/releases/tag/v2.0.0).
> 
> It also updates Babeltrace 1 to [1.5.11](https://github.com/efficios/babeltrace/releases/tag/v1.5.11) (which mainly improves compatibility with newer Python and dependency versions) and adds a derivation for its Python bindings as well.
> 
> Cross-compilation (targeting aarch64-linux on x86_64-linux) has been tested.

My additions to that are:
- update babeltrace2 from 2.0.5 to 2.0.6
- fix/patch both babeltrace python packages to work with Python 3.12.
- run `nixpkgs-review`

Closes #285681 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

hacker1024 committed:
python311Packages.babeltrace: init at 1.5.11

wentasah commented:
Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.perf</li>
    <li>linuxKernel.packages.linux_4_19.perf.debug</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf.debug</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>rocmPackages_5.mivisionx (rocmPackages_5.mivisionx-hip)</li>
    <li>rocmPackages_5.mivisionx-cpu</li>
    <li>rocmPackages_5.mivisionx-opencl</li>
  </ul>
</details>
<details>
  <summary>57 packages built:</summary>
  <ul>
    <li>babeltrace</li>
    <li>babeltrace2</li>
    <li>cargo-flamegraph</li>
    <li>ceph</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev</li>
    <li>ceph.doc</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man</li>
    <li>hotspot</li>
    <li>linuxKernel.packages.linux_5_10.perf</li>
    <li>linuxKernel.packages.linux_5_10.perf.debug</li>
    <li>linuxKernel.packages.linux_5_10_hardened.perf</li>
    <li>linuxKernel.packages.linux_5_10_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_5_15.perf</li>
    <li>linuxKernel.packages.linux_5_15.perf.debug</li>
    <li>linuxKernel.packages.linux_5_15_hardened.perf</li>
    <li>linuxKernel.packages.linux_5_15_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_5_4.perf</li>
    <li>linuxKernel.packages.linux_5_4.perf.debug</li>
    <li>linuxKernel.packages.linux_5_4_hardened.perf</li>
    <li>linuxKernel.packages.linux_5_4_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_6_1.perf</li>
    <li>linuxKernel.packages.linux_6_1.perf.debug</li>
    <li>linuxKernel.packages.linux_6_1_hardened.perf</li>
    <li>linuxKernel.packages.linux_6_1_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_6_6.perf</li>
    <li>linuxKernel.packages.linux_6_6.perf.debug</li>
    <li>linuxKernel.packages.linux_hardened.perf (linuxKernel.packages.linux_6_6_hardened.perf)</li>
    <li>linuxKernel.packages.linux_hardened.perf.debug (linuxKernel.packages.linux_6_6_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_8.perf</li>
    <li>linuxKernel.packages.linux_6_8.perf.debug</li>
    <li>linuxKernel.packages.linux_latest_libre.perf</li>
    <li>linuxKernel.packages.linux_latest_libre.perf.debug</li>
    <li>linuxKernel.packages.linux_libre.perf</li>
    <li>linuxKernel.packages.linux_libre.perf.debug</li>
    <li>linuxKernel.packages.linux_lqx.perf</li>
    <li>linuxKernel.packages.linux_lqx.perf.debug</li>
    <li>linuxKernel.packages.linux_xanmod.perf</li>
    <li>linuxKernel.packages.linux_xanmod.perf.debug</li>
    <li>linuxKernel.packages.linux_xanmod_latest.perf (linuxKernel.packages.linux_xanmod_stable.perf)</li>
    <li>linuxKernel.packages.linux_xanmod_latest.perf.debug (linuxKernel.packages.linux_xanmod_stable.perf.debug)</li>
    <li>linuxKernel.packages.linux_zen.perf</li>
    <li>linuxKernel.packages.linux_zen.perf.debug</li>
    <li>lttv</li>
    <li>python311Packages.babeltrace</li>
    <li>python311Packages.babeltrace2</li>
    <li>python312Packages.babeltrace</li>
    <li>python312Packages.babeltrace2</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qtcreator</li>
    <li>samba4Full</li>
    <li>samba4Full.dev</li>
    <li>samba4Full.man</li>
  </ul>
</details>

wentasah commented:
`rocmPackages_5.mivisionx` fails due to failure of its dependency `python311Packages.torchWithRocm`, which [fails on master since update to 2.3.0 few days ago](https://hydra.nixos.org/build/259580033). Therefore I think this can be safely merged.

doronbehar commented on pkgs/development/tools/misc/babeltrace/common.nix:
```diff
@@ -0,0 +1,43 @@
+, pkg-config
+, glib
+, elfutils
+
```
I'd not separate the `pname` and `version` from the other arguments...

wentasah replied:
Yes, this simplify things.

doronbehar commented on pkgs/development/tools/misc/babeltrace/common.nix:
```diff
@@ -0,0 +1,43 @@
+  inherit version;
+
+  src = fetchzip {
+    url = "https://www.efficios.com/files/babeltrace/${finalAttrs.pname}-${finalAttrs.version}.tar.bz2";
```
And then you'd be able to use `pname` and `version` without the `finalAttrs.` prefix...

wentasah replied:
I removed `finalAttrs` completely, because it's  not used for anything.

doronbehar commented on pkgs/development/tools/misc/babeltrace/common.nix:
```diff
@@ -0,0 +1,43 @@
+  # In 2.x, there is no pre-generated ./configure script.
+  # Re-generate with nixpkgs's autoconf. This requires glib to be present in
+  # nativeBuildInputs for its m4 macros to be present.
+  nativeBuildInputs = args.nativeBuildInputs or [ ] ++ [ autoreconfHook pkg-config glib ];
```
Then here you'd use an argument I'd name `extraNativeBuildInputs`.

doronbehar commented on pkgs/development/tools/misc/babeltrace/common.nix:
```diff
@@ -0,0 +1,43 @@
+  # nativeBuildInputs for its m4 macros to be present.
+  nativeBuildInputs = args.nativeBuildInputs or [ ] ++ [ autoreconfHook pkg-config glib ];
+
+  buildInputs = args.buildInputs or [ ] ++ [ glib elfutils ];
```
Same here - use `extraBuildInputs`.

wentasah replied:
This was actually not used at all so I removed it.

doronbehar commented on pkgs/development/python-modules/babeltrace2/default.nix:
```diff
@@ -0,0 +1,17 @@
+, swig
+}:
+
+toPythonModule (babeltrace2.overrideAttrs ({ nativeBuildInputs ? [ ], configureFlags ? [ ], ... }: {
```
IMO the interface you used in the first commits is not worth it if eventually you use `overrideAttrs`.

wentasah replied:
The interface is for covering the differences between v1 and v2. Overriding is used for changes in Python package builds. If we extend the interface from the first commit to also allow conditional building of Python packages (e.g. by adding `enablePython` as you suggest bellow), this would complicate the expression a lot. I find it more manageable to use overrides here.

doronbehar replied:
> The interface is for covering the differences between v1 and v2. Overriding is used for changes in Python package builds. If we extend the interface from the first commit to also allow conditional building of Python packages (e.g. by adding `enablePython` as you suggest bellow), this would complicate the expression a lot. I find it more manageable to use overrides here.

Personally it feels more complicated to me to use 2 different overriding schemes, and scattering the logic between different files. After all both derivations are built from the same source. I also find it very confusing the `nativeBuildInputs ? []` and `configureFlags ? []` function... What is the value given by reading these? 

If you hadn't had used a function with two attribute sets as arguments it would have been at least more readable.

wentasah replied:
> I also find it very confusing the `nativeBuildInputs ? []` and `configureFlags ? []` function...

Yeah, I thought that this is a bit awkward. I'll try to come up with a cleaner solution. Perhaps splitting v1 and v2 as you suggest.

doronbehar commented on pkgs/development/python-modules/babeltrace/default.nix:
```diff
@@ -0,0 +1,14 @@
+toPythonModule (babeltrace.overrideAttrs ({ nativeBuildInputs ? [ ], configureFlags ? [ ], ... }: {
+  nativeBuildInputs = nativeBuildInputs ++ [ swig2 ];
+
+  configureFlags = configureFlags ++ [ "--enable-python-bindings" ];
```
Why not put this logic in an `enablePython` flag?

wentasah replied:
See above.

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18118,9 +18118,11 @@ with pkgs;
   b4 = callPackage ../development/tools/b4 { };
 
-  babeltrace = callPackage ../development/tools/misc/babeltrace/v1.nix { };
+  babeltracePackages = callPackage ../development/tools/misc/babeltrace { };
```
This is too much IMO..

wentasah replied:
This was recommended [here](https://github.com/NixOS/nixpkgs/pull/285681#issuecomment-1922860408). It think that using the recommended `babeltraceVersions` would be better than current `babeltracePackages` so I changed it.

doronbehar replied:
> This was recommended [here](https://github.com/NixOS/nixpkgs/pull/285681#issuecomment-1922860408). It think that using the recommended `babeltraceVersions` would be better than current `babeltracePackages` so I changed it.

I liked most this option:

> * Use something like `inherit (callPackages ...) babeltrace babeltrace 2`

I think there is no value in a `babeltracePackages` attribute set, when the attributes within this attribute set don't interact with each other, and there are just 2 of them. Also note it creates a bit more noise when you `nix search`.

doronbehar commented on pkgs/development/python-modules/babeltrace/default.nix:
```diff
@@ -9,6 +9,10 @@ toPythonModule (babeltrace.overrideAttrs ({ nativeBuildInputs ? [ ], configureFl
+  # Nix treats nativeBuildInputs specially for cross-compilation, but in this
+  # case, cross-compilation is accounted for explicitly. Using the variables
+  # ensures that the platform setup isn't messed with further. It also allows
+  # regular Python to be added in the future if it is ever needed.
```
This comment should be fixed up to the `python3Packages.` commits.

doronbehar commented on pkgs/development/python-modules/babeltrace/default.nix:
```diff
@@ -2,10 +2,11 @@
 
 toPythonModule (babeltrace.overrideAttrs ({ nativeBuildInputs ? [ ], configureFlags ? [ ], ... }: {
-  nativeBuildInputs = nativeBuildInputs ++ [ swig2 ];
+  nativeBuildInputs = nativeBuildInputs ++ [ swig2 setuptools ];
```
Fix this up to the `python312Packages.babeltrace: init` commits.

doronbehar commented on pkgs/development/python-modules/babeltrace2/default.nix:
```diff
@@ -2,10 +2,32 @@
+  patches = patches ++ [
+    # Patches needed for Python 3.12
+    (fetchpatch {
+      # python: Use standalone 'sysconfig' module
```
These patches too should be squashed to the `init` commits..

doronbehar reviewed (commented):
P.S If you'd use `python311Packages.babeltrace` and not `python3Packages.babeltrace` in the commit prefixes, you'd convince ofborg to build this attributes for you automatically. (The `python3Packages` prefixed attributes are not discovered by Hydra and hence are not discoved by ofborg as well).

doronbehar commented:
Sorry it took so long to discover this PR.. It looks good overall.

wentasah commented:
So I addressed most of the @doronbehar's comments including squashing commits together and updating commit titles to trigger ofborg. Now, I'm running `nixpkgs-review`. I'll post the results once it finishes.

wentasah commented:
Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.perf</li>
    <li>linuxKernel.packages.linux_4_19.perf.debug</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf.debug</li>
    <li>linuxPackages_4_19_hardened.perf</li>
    <li>linuxPackages_4_19_hardened.perf.debug</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx</li>
    <li>rocmPackages_5.mivisionx-cpu</li>
    <li>rocmPackages_5.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx-opencl</li>
    <li>rocmPackages_6.mivisionx</li>
    <li>rocmPackages_6.mivisionx-cpu</li>
    <li>rocmPackages_6.mivisionx-hip</li>
  </ul>
</details>
<details>
  <summary>60 packages built:</summary>
  <ul>
    <li>babeltrace</li>
    <li>babeltrace2</li>
    <li>cargo-flamegraph</li>
    <li>ceph (ceph-dev)</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev (ceph-dev.dev)</li>
    <li>ceph.doc (ceph-dev.doc)</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man (ceph-dev.man)</li>
    <li>hotspot</li>
    <li>linuxPackages_5_10_hardened.perf (linuxKernel.packages.linux_5_10.perf)</li>
    <li>linuxPackages_5_10_hardened.perf.debug (linuxKernel.packages.linux_5_10.perf.debug)</li>
    <li>linuxPackages_5_15_hardened.perf (linuxKernel.packages.linux_5_15.perf)</li>
    <li>linuxPackages_5_15_hardened.perf.debug (linuxKernel.packages.linux_5_15.perf.debug)</li>
    <li>linuxKernel.packages.linux_5_4.perf</li>
    <li>linuxKernel.packages.linux_5_4.perf.debug</li>
    <li>linuxPackages_5_4_hardened.perf (linuxKernel.packages.linux_5_4_hardened.perf)</li>
    <li>linuxPackages_5_4_hardened.perf.debug (linuxKernel.packages.linux_5_4_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_1.perf</li>
    <li>linuxKernel.packages.linux_6_1.perf.debug</li>
    <li>linuxPackages_6_1_hardened.perf (linuxKernel.packages.linux_6_1_hardened.perf)</li>
    <li>linuxPackages_6_1_hardened.perf.debug (linuxKernel.packages.linux_6_1_hardened.perf.debug)</li>
    <li>linuxPackages.perf (linuxKernel.packages.linux_6_6.perf)</li>
    <li>linuxPackages.perf.debug (linuxKernel.packages.linux_6_6.perf.debug)</li>
    <li>linuxPackages_hardened.perf (linuxPackages_6_6_hardened.perf)</li>
    <li>linuxPackages_hardened.perf.debug (linuxPackages_6_6_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_8.perf</li>
    <li>linuxKernel.packages.linux_6_8.perf.debug</li>
    <li>linuxPackages_6_8_hardened.perf (linuxKernel.packages.linux_6_8_hardened.perf)</li>
    <li>linuxPackages_6_8_hardened.perf.debug (linuxKernel.packages.linux_6_8_hardened.perf.debug)</li>
    <li>linuxPackages_latest.perf (linuxKernel.packages.linux_6_9.perf)</li>
    <li>linuxPackages_latest.perf.debug (linuxKernel.packages.linux_6_9.perf.debug)</li>
    <li>linuxPackages_6_9_hardened.perf (linuxKernel.packages.linux_6_9_hardened.perf)</li>
    <li>linuxPackages_6_9_hardened.perf.debug (linuxKernel.packages.linux_6_9_hardened.perf.debug)</li>
    <li>linuxPackages_latest-libre.perf (linuxKernel.packages.linux_latest_libre.perf)</li>
    <li>linuxPackages_latest-libre.perf.debug (linuxKernel.packages.linux_latest_libre.perf.debug)</li>
    <li>linuxPackages-libre.perf (linuxKernel.packages.linux_libre.perf)</li>
    <li>linuxPackages-libre.perf.debug (linuxKernel.packages.linux_libre.perf.debug)</li>
    <li>linuxPackages_lqx.perf (linuxKernel.packages.linux_lqx.perf)</li>
    <li>linuxPackages_lqx.perf.debug (linuxKernel.packages.linux_lqx.perf.debug)</li>
    <li>linuxPackages_xanmod.perf (linuxKernel.packages.linux_xanmod.perf)</li>
    <li>linuxPackages_xanmod.perf.debug (linuxKernel.packages.linux_xanmod.perf.debug)</li>
    <li>linuxPackages_xanmod_latest.perf (linuxKernel.packages.linux_xanmod_latest.perf ,linuxPackages_xanmod_stable.perf)</li>
    <li>linuxPackages_xanmod_latest.perf.debug (linuxKernel.packages.linux_xanmod_latest.perf.debug ,linuxPackages_xanmod_stable.perf.debug)</li>
    <li>linuxPackages_zen.perf (linuxKernel.packages.linux_zen.perf)</li>
    <li>linuxPackages_zen.perf.debug (linuxKernel.packages.linux_zen.perf.debug)</li>
    <li>lttv</li>
    <li>python311Packages.babeltrace</li>
    <li>python311Packages.babeltrace2</li>
    <li>python312Packages.babeltrace</li>
    <li>python312Packages.babeltrace2</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qtcreator</li>
    <li>rocmPackages.rocgdb (rocmPackages_6.rocgdb)</li>
    <li>sambaFull (samba4Full)</li>
    <li>sambaFull.dev (samba4Full.dev)</li>
    <li>sambaFull.man (samba4Full.man)</li>
  </ul>
</details>

doronbehar commented:
I'm also not sure that the right thing to do here is to separate the expressions for the v1 and version 2 attributes... If they are separate, they might eventually change through out time differently. If the current `common.nix` would have been very complicated, I'd agree it would be smarter to share a `common.nix` file, like what they do with the different Thunderbird variants, and the chromium and Firefox variants... This is such a simple and classic expression that it doesn't worth it IMO.

wentasah committed:
babeltrace: add myself as maintainer

wentasah committed:
babeltrace: add update script

wentasah committed:
python311Packages.babeltrace2: init at 2.0.6

A new attribute overriding babeltrace2 is created, rather than enabling
Python support in babeltrace2 itself, in order to reduce the closure
size and leave babeltrace2 Python-version-agnostic.

Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

wentasah committed:
babeltrace2: init at 2.0.6

Separate from babeltrace 1.x as "the Babeltrace 2 project is completely
independent from Babeltrace 1".

https://github.com/efficios/babeltrace/releases/tag/v2.0.0

Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

wentasah commented:
I rebased to nixos-unstable and separated v1 and v2 to their own independent packages. The result is really simpler.

@bjornfor I added myself as co-maintainer to v1. Do you agree?

Other changes:
- added an update script to make updates easier/automated
- v2 uses `enablePython` parameter (I'll probably do the same with v1 - sorry for not doing it in one shot)

wentasah commented:
Now, even v1 uses `enablePython`.

@hacker1024 It seems that cross-compilation to aarch64 was important for you. Cross-compilation works with Python 3.11, but fails with 3.12. Is it ok for you?

bjornfor commented:
> @bjornfor I added myself as co-maintainer to v1. Do you agree?

Yes. Thanks!

wentasah commented:
Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.perf</li>
    <li>linuxKernel.packages.linux_4_19.perf.debug</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf.debug</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx</li>
    <li>rocmPackages_5.mivisionx-cpu</li>
    <li>rocmPackages_5.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx-opencl</li>
  </ul>
</details>
<details>
  <summary>60 packages built:</summary>
  <ul>
    <li>babeltrace</li>
    <li>babeltrace2</li>
    <li>cargo-flamegraph</li>
    <li>ceph</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev</li>
    <li>ceph.doc</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man</li>
    <li>hotspot</li>
    <li>linuxKernel.packages.linux_5_10.perf (linuxKernel.packages.linux_5_10_hardened.perf)</li>
    <li>linuxKernel.packages.linux_5_10.perf.debug (linuxKernel.packages.linux_5_10_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_5_15.perf (linuxKernel.packages.linux_5_15_hardened.perf)</li>
    <li>linuxKernel.packages.linux_5_15.perf.debug (linuxKernel.packages.linux_5_15_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_5_4.perf</li>
    <li>linuxKernel.packages.linux_5_4.perf.debug</li>
    <li>linuxKernel.packages.linux_5_4_hardened.perf</li>
    <li>linuxKernel.packages.linux_5_4_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_6_1.perf</li>
    <li>linuxKernel.packages.linux_6_1.perf.debug</li>
    <li>linuxKernel.packages.linux_6_1_hardened.perf</li>
    <li>linuxKernel.packages.linux_6_1_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_6_6.perf</li>
    <li>linuxKernel.packages.linux_6_6.perf.debug</li>
    <li>linuxKernel.packages.linux_hardened.perf (linuxKernel.packages.linux_6_6_hardened.perf)</li>
    <li>linuxKernel.packages.linux_hardened.perf.debug (linuxKernel.packages.linux_6_6_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_8.perf</li>
    <li>linuxKernel.packages.linux_6_8.perf.debug</li>
    <li>linuxKernel.packages.linux_6_8_hardened.perf</li>
    <li>linuxKernel.packages.linux_6_8_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_6_9.perf</li>
    <li>linuxKernel.packages.linux_6_9.perf.debug</li>
    <li>linuxKernel.packages.linux_6_9_hardened.perf</li>
    <li>linuxKernel.packages.linux_6_9_hardened.perf.debug</li>
    <li>linuxKernel.packages.linux_latest_libre.perf</li>
    <li>linuxKernel.packages.linux_latest_libre.perf.debug</li>
    <li>linuxKernel.packages.linux_libre.perf</li>
    <li>linuxKernel.packages.linux_libre.perf.debug</li>
    <li>linuxKernel.packages.linux_lqx.perf</li>
    <li>linuxKernel.packages.linux_lqx.perf.debug</li>
    <li>linuxKernel.packages.linux_xanmod.perf</li>
    <li>linuxKernel.packages.linux_xanmod.perf.debug</li>
    <li>linuxKernel.packages.linux_xanmod_latest.perf (linuxKernel.packages.linux_xanmod_stable.perf)</li>
    <li>linuxKernel.packages.linux_xanmod_latest.perf.debug (linuxKernel.packages.linux_xanmod_stable.perf.debug)</li>
    <li>linuxKernel.packages.linux_zen.perf</li>
    <li>linuxKernel.packages.linux_zen.perf.debug</li>
    <li>lttv</li>
    <li>python311Packages.babeltrace</li>
    <li>python311Packages.babeltrace2</li>
    <li>python312Packages.babeltrace</li>
    <li>python312Packages.babeltrace2</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qtcreator</li>
    <li>rocmPackages.rocgdb</li>
    <li>samba4Full</li>
    <li>samba4Full.dev</li>
    <li>samba4Full.man</li>
  </ul>
</details>

hacker1024 commented:
> Now, even v1 uses `enablePython`.
> 
> @hacker1024 It seems that cross-compilation to aarch64 was important for you. Cross-compilation works with Python 3.11, but fails with 3.12. Is it ok for you?

That's fine for me for now, I'm happy to look into it myself for 3.12 later. I'll probably need it once we switch to [ROS Jazzy Jalisco with Python 3.12](https://www.ros.org/reps/rep-2000.html#jazzy-jalisco-may-2024-may-2029).

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+    hash = "sha256-L4YTqPxvWynUBnmAQnlJ2RNbEv9MhBxQOsqbWix8ZwU=";
+  };
+
+  patches = lib.optionals enablePython [
+    # Patches needed for Python 3.12
+    (fetchpatch {
+      # python: Use standalone 'sysconfig' module
+      url = "https://github.com/efficios/babeltrace/commit/452480eb6820df9973d50431a479ca547815ae08.patch";
+      hash = "sha256-YgUKHJzdliNUsTY29E0xxcUjqVWn4EvxyTs0B+O+jrI=";
+    })
+    (fetchpatch {
+      # python: replace distutils with setuptools
+      url = "https://github.com/efficios/babeltrace/commit/6ec97181a525a3cd64cedbcd0df905ed9e84ba03.patch";
+      hash = "sha256-1hlEkPcRUpf2+iEXqHXcCDOaLTg+eaVcahqZlA8m5QY=";
+    })
+    (fetchpatch {
+      # fix: python: monkey patch the proper sysconfig implementation
+      url = "https://github.com/efficios/babeltrace/commit/927263e4ea62877af7240cfdb1514ae949dbfc2e.patch";
+      hash = "sha256-HNRQ7uw26FUKCQ/my6//OL2xsHdOGlQUq5zIKtg9OGw=";
+    })
+  ];
```
I'm pretty sure that if you'd apply these patches for the build with `enablePython = false`,  it wouldn't have hurt it. I find it slightly better to do apply them unconditionally because this way if all of a sudden they won't apply, all consumers of this expression - i.e consumers of all babeltrace2 attributes will notice. Think also of `nixpkgs-update` that will not fail to update `babeltrace2` even though the patches don't apply, because the current conditional doesn't apply them.

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, fetchpatch
+, autoreconfHook
+, pkg-config
+, glib
+, elfutils
+, bison
+, flex
+, enableMan ? true, asciidoc, xmlto, docbook_xml_dtd_45, docbook_xsl
+, enablePython ? false, python ? null, pythonPackages ? null, swig, ensureNewerSourcesForZipFilesHook
+}:
```
Could you please run `nixfmt` from `nixfmt-rfc-style` on this please? :pray:

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+    })
+  ];
+
+  nativeBuildInputs =
+    [ autoreconfHook pkg-config glib bison flex ]
+    ++ lib.optionals enableMan [ asciidoc xmlto docbook_xml_dtd_45 docbook_xsl ]
+    ++ lib.optionals enablePython [ swig pythonPackages.setuptools ensureNewerSourcesForZipFilesHook ];
+
+  buildInputs =
+    [ glib elfutils ]
+    ++ lib.optional enablePython python;
```
Here the formatting is awful IMO, and `nixfmt` will fix it..

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+    (lib.enableFeature enableMan "man-pages")
+    (lib.enableFeature enablePython "python-bindings")
+    (lib.enableFeature enablePython "python-plugins")
+  ] ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) "--disable-debug-info";
```
Why not use `lib.enableFeature` for this one as well?

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+    ++ lib.optional enablePython python;
+
+  configureFlags = [
+    (lib.enableFeature enableMan "man-pages")
```
Do you think we can unconditionally enable man pages and use `outputs = [ "out" "man" "dev" ];` so they won't take too much space in closure if someone doesn't want them?

doronbehar commented on pkgs/by-name/ba/babeltrace2/package.nix:
```diff
@@ -0,0 +1,73 @@
+  ] ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) "--disable-debug-info";
+
+
+  makeFlags = [ "CFLAGS=-Wno-error=stringop-truncation -Wno-error=null-dereference" ];
```
Just out of curiosity, is upstream aware that they need these loosening CFLAGS to make it work? It'd be nice to put a link if they do.

wentasah replied:
I don't see any upstream commit addressing that. I'll let them know later, I'm in hurry right know. I just put there a comment that its needed when cross-compiling Python bindings.

doronbehar commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1336,6 +1336,13 @@ self: super: with self; {
+    inherit (self) python;
+    pythonPackages = self;
+    enablePython = true;
+    enableMan = false;
```
Why not enabling man pages?

doronbehar commented on pkgs/development/tools/misc/babeltrace/default.nix:
```diff
@@ -2,11 +2,11 @@
 
 stdenv.mkDerivation rec {
```
After the 1.5.8 -> 1.5.11 commit, or even before it, it'd be nice to format with `nixfmt` the expression, and also switch to `finalAttrs:` pattern.

wentasah replied:
I reformatted it as last commit (no time for resolving conflicts right now :-). 

Also, is there any advantage of adding the `finalAttrs` parameter if it is not used in the actual Nix expression? I don't think so and I believe that it makes evaluation a bit slower.

wentasah replied:
Here we could replace `rec` and `version` with `finalAttrs`, but it doesn't help much, because we would need to override the whole `src` due to `hash` change. Or am I missing something?

doronbehar commented on pkgs/development/tools/misc/babeltrace/default.nix:
```diff
@@ -31,7 +31,7 @@ stdenv.mkDerivation rec {
     homepage = "https://www.efficios.com/babeltrace";
     license = licenses.mit;
     platforms = platforms.linux;
-    maintainers = [ maintainers.bjornfor ];
```
Also, there is a common claim among Nixpkgs contributors that it is better to not use `with lib;` for the whole `meta` scope, and use `lib.licenses.`, `lib.platforms.` and `with lib.maintainers;` here - that also could be amended to the `nixfmt` commit of `babeltrace`.

doronbehar reviewed (commented):
Looks much better!

doronbehar commented:
> Cross-compilation works with Python 3.11, but fails with 3.12. Is it ok for you?

Could you share please how it fails? Maybe I can help.

siraben commented:
Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.perf</li>
    <li>linuxKernel.packages.linux_4_19.perf.debug</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf</li>
    <li>linuxKernel.packages.linux_4_19_hardened.perf.debug</li>
    <li>linuxPackages_4_19_hardened.perf</li>
    <li>linuxPackages_4_19_hardened.perf.debug</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx</li>
    <li>rocmPackages_5.mivisionx-cpu</li>
    <li>rocmPackages_5.mivisionx-hip</li>
    <li>rocmPackages_5.mivisionx-opencl</li>
    <li>rocmPackages_6.mivisionx</li>
    <li>rocmPackages_6.mivisionx-cpu</li>
    <li>rocmPackages_6.mivisionx-hip</li>
  </ul>
</details>
<details>
  <summary>13 packages failed to build:</summary>
  <ul>
    <li>ceph (ceph-dev)</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev (ceph-dev.dev)</li>
    <li>ceph.doc (ceph-dev.doc)</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man (ceph-dev.man)</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>sambaFull (samba4Full)</li>
    <li>sambaFull.dev (samba4Full.dev)</li>
    <li>sambaFull.man (samba4Full.man)</li>
  </ul>
</details>
<details>
  <summary>47 packages built:</summary>
  <ul>
    <li>babeltrace</li>
    <li>babeltrace2</li>
    <li>cargo-flamegraph</li>
    <li>hotspot</li>
    <li>linuxPackages_5_10_hardened.perf (linuxKernel.packages.linux_5_10.perf)</li>
    <li>linuxPackages_5_10_hardened.perf.debug (linuxKernel.packages.linux_5_10.perf.debug)</li>
    <li>linuxPackages_5_15_hardened.perf (linuxKernel.packages.linux_5_15.perf)</li>
    <li>linuxPackages_5_15_hardened.perf.debug (linuxKernel.packages.linux_5_15.perf.debug)</li>
    <li>linuxKernel.packages.linux_5_4.perf</li>
    <li>linuxKernel.packages.linux_5_4.perf.debug</li>
    <li>linuxPackages_5_4_hardened.perf (linuxKernel.packages.linux_5_4_hardened.perf)</li>
    <li>linuxPackages_5_4_hardened.perf.debug (linuxKernel.packages.linux_5_4_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_1.perf</li>
    <li>linuxKernel.packages.linux_6_1.perf.debug</li>
    <li>linuxPackages_6_1_hardened.perf (linuxKernel.packages.linux_6_1_hardened.perf)</li>
    <li>linuxPackages_6_1_hardened.perf.debug (linuxKernel.packages.linux_6_1_hardened.perf.debug)</li>
    <li>linuxPackages.perf (linuxKernel.packages.linux_6_6.perf)</li>
    <li>linuxPackages.perf.debug (linuxKernel.packages.linux_6_6.perf.debug)</li>
    <li>linuxPackages_hardened.perf (linuxPackages_6_6_hardened.perf)</li>
    <li>linuxPackages_hardened.perf.debug (linuxPackages_6_6_hardened.perf.debug)</li>
    <li>linuxKernel.packages.linux_6_8.perf</li>
    <li>linuxKernel.packages.linux_6_8.perf.debug</li>
    <li>linuxPackages_6_8_hardened.perf (linuxKernel.packages.linux_6_8_hardened.perf)</li>
    <li>linuxPackages_6_8_hardened.perf.debug (linuxKernel.packages.linux_6_8_hardened.perf.debug)</li>
    <li>linuxPackages_latest.perf (linuxKernel.packages.linux_6_9.perf)</li>
    <li>linuxPackages_latest.perf.debug (linuxKernel.packages.linux_6_9.perf.debug)</li>
    <li>linuxPackages_6_9_hardened.perf (linuxKernel.packages.linux_6_9_hardened.perf)</li>
    <li>linuxPackages_6_9_hardened.perf.debug (linuxKernel.packages.linux_6_9_hardened.perf.debug)</li>
    <li>linuxPackages_latest-libre.perf (linuxKernel.packages.linux_latest_libre.perf)</li>
    <li>linuxPackages_latest-libre.perf.debug (linuxKernel.packages.linux_latest_libre.perf.debug)</li>
    <li>linuxPackages-libre.perf (linuxKernel.packages.linux_libre.perf)</li>
    <li>linuxPackages-libre.perf.debug (linuxKernel.packages.linux_libre.perf.debug)</li>
    <li>linuxPackages_lqx.perf (linuxKernel.packages.linux_lqx.perf)</li>
    <li>linuxPackages_lqx.perf.debug (linuxKernel.packages.linux_lqx.perf.debug)</li>
    <li>linuxPackages_xanmod.perf (linuxKernel.packages.linux_xanmod.perf)</li>
    <li>linuxPackages_xanmod.perf.debug (linuxKernel.packages.linux_xanmod.perf.debug)</li>
    <li>linuxPackages_xanmod_latest.perf (linuxKernel.packages.linux_xanmod_latest.perf ,linuxPackages_xanmod_stable.perf)</li>
    <li>linuxPackages_xanmod_latest.perf.debug (linuxKernel.packages.linux_xanmod_latest.perf.debug ,linuxPackages_xanmod_stable.perf.debug)</li>
    <li>linuxPackages_zen.perf (linuxKernel.packages.linux_zen.perf)</li>
    <li>linuxPackages_zen.perf.debug (linuxKernel.packages.linux_zen.perf.debug)</li>
    <li>lttv</li>
    <li>python311Packages.babeltrace</li>
    <li>python311Packages.babeltrace2</li>
    <li>python312Packages.babeltrace</li>
    <li>python312Packages.babeltrace2</li>
    <li>qtcreator</li>
    <li>rocmPackages.rocgdb (rocmPackages_6.rocgdb)</li>
  </ul>
</details>

wentasah commented:
> > Cross-compilation works with Python 3.11, but fails with 3.12. Is it ok for you?
> 
> Could you share please how it fails? Maybe I can help.

It fails when cross-compiling Python itself:
```
python3> In file included from ./Include/internal/pycore_object.h:13,
python3>                  from Objects/cellobject.c:4:
python3> ./Include/internal/pycore_interp.h:193:24: error: field 'dtoa' has incomplete type
python3>   193 |     struct _dtoa_state dtoa;
python3>       |                        ^~~~
```

doronbehar commented:
> It fails when cross-compiling Python itself:
> 
> ```
> python3> In file included from ./Include/internal/pycore_object.h:13,
> python3>                  from Objects/cellobject.c:4:
> python3> ./Include/internal/pycore_interp.h:193:24: error: field 'dtoa' has incomplete type
> python3>   193 |     struct _dtoa_state dtoa;
> python3>       |                        ^~~~
> ```

Could you share in a gist the full build log please?

wentasah commented:
[Full python build log](https://gist.github.com/wentasah/d6358baa0fb4eb36dbfdedfc702cd968) as produced by `nix build .#pkgsCross.aarch64-multiplatform.python312Packages.babeltrace`

wentasah commented:
> Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 16 packages marked as broken and skipped:
> 13 packages failed to build:

Probably unrelated. Now I see the same, yesterday it was OK.

doronbehar commented:
> [Full python build log](https://gist.github.com/wentasah/d6358baa0fb4eb36dbfdedfc702cd968) as produced by `nix build .#pkgsCross.aarch64-multiplatform.python312Packages.python`

cc @FRidh .

doronbehar commented:
> > Result of `nixpkgs-review pr 310617` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> > 16 packages marked as broken and skipped:
> > 13 packages failed to build:
> 
> Probably unrelated. Now I see the same, yesterday it was OK.

Yes it could be failures related to the processing power of the machine you used. @siraben try to use `nix build --max-jobs 1 --cores 2` on those failing attributes.

siraben commented:
@doronbehar Ok, rerunning.

wentasah committed:
babeltrace: reformat with nixfmt-rfc-style

- Don't use `meta = with lib;`.
- Use (finalAttrs:{}) pattern.
- Use lib.enableFeature for debug-info ./configure flag.
- Use pkgs/by-name

wentasah commented:
Addressed most of @doronbehar's comments.

doronbehar committed:
babeltrace: 1.5.8 -> 1.5.11

doronbehar commented:
> I reformatted it as last commit (no time for resolving conflicts right now :-).

Indeed the rebase wasn't trivial :) Now it is done.

> Also, is there any advantage of adding the `finalAttrs` parameter if it is not used in the actual Nix expression? I don't think so and I believe that it makes evaluation a bit slower.

Yes. Using `rec` is slower then evaluating a function.

> Addressed most of @doronbehar's comments.

So I hadn't had the patience to wait for it to be perfect, so I did it all myself. Your final approval on my changes will be great, if not, I'll simply wait for CI.

wentasah commented:
> Yes. Using `rec` is slower then evaluating a function.

Great. I was not aware of that.

> So I hadn't had the patience to wait for it to be perfect, so I did it all myself. Your final approval on my changes will be great, if not, I'll simply wait for CI.

Thank you for finishing and merging this. I was traveling with my family - sorry for not having time to give it a final testing.
